### PR TITLE
Speed up projections of structural records

### PR DESCRIPTION
**Pull request by hurryabit**
Currently, for every projection of a structural record, we first do a
linear search through the field names to find the index of the field
value in the record value. Then we use this index to get the actual
value. However, since we started sorting the fields in alphabetic order
both in types and in values, this lookup will always yield the same
result for a fixed record type and a fixed field name, regardless of the
value of the record.

This PR changes the AST node for structural record projection to
contain the index of the field within the record as well. This
information is not contained in the DAML-LF archive itself and must
hence be filled in by the DAML-LF type checker.

The only context in which we use structural record prejections are
typeclass method invocations, which do a little bit more than just the
projection. Unfortunately, typeclass method invocations are the small
unit of work I can benchmark using the setup I have. In the end, they
are also the feature our users really care about.

My benchmarks show that the time a single typeclass method invocation
takes has dropped from ca. 210ns to ca. 130ns. That's a speedup of
ca. 1.6x.

CHANGELOG_BEGIN
[DAML Interpreter] Typeclass method invocation is now 1.6x faster.
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

---
This change is [Reviewable](https://reviewable.io/reviews/digital-asset/daml/7740)

### CODE REVIEW

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala [hunk 1]
```diff
@@ -84,8 +84,8 @@ object Ast {
   /** Struct construction. */
   final case class EStructCon(fields: ImmArray[(FieldName, Expr)]) extends Expr
 
-  /** Struct projection. */
-  final case class EStructProj(field: FieldName, struct: Expr) extends Expr
+  /** Struct projection. The value for `fieldIndex` is fillen in by the type checker. */
+  final case class EStructProj(field: FieldName, var fieldIndex: Option[Int], struct: Expr) extends Expr
```
Comment (remyhaemmerle-da) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala:88]:
```suggestion
  final case class EStructProj(field: FieldName, struct: Expr) extends Expr {
    private[lf] var fieldIndex: Option[Int] = None
  }
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 2]
```diff
@@ -374,8 +374,11 @@ private[lf] final class Compiler(
           SEBuiltin(SBStructCon(fieldsInputOrder)),
           fields.iterator.map { case (_, e) => compile(e) }.toArray
         )
-      case EStructProj(field, struct) =>
-        SBStructProj(field)(compile(struct))
+      case EStructProj(field @ _, fieldIndex, struct) =>
+        fieldIndex match {
+          case None => throw CompilationError(s"structural record projection for field ${field} has no index")
+          case Some(index) => SBStructProj(index)(compile(struct))
+        }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:381]:
```suggestion
      case strucProj: EStructProj =>
        strucProj.fieldIndex match {
          case None => throw CompilationError(s"structural record projection for field ${field} has no index")
          case Some(index) => SBStructProj(index)(compile(strucProj.struct))
        }
```

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 3]
```diff
@@ -835,6 +835,7 @@ private[archive] class DecodeV1(minor: LV.Minor) extends Decode.OfPackage[PLF.Pa
               structProj.getFieldInternedStr,
               "Expr.StructProj.field.field"
             ),
+            fieldIndex = None,  // This will be filled in by the type checker.
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:838]:
```suggestion
```

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/ExprTraversable.scala [hunk 4]
```diff
@@ -27,7 +27,7 @@ private[validation] object ExprTraversable {
         f(arg)
       case EStructCon(fields) =>
         fields.values.foreach(f)
-      case EStructProj(field @ _, struct) =>
+      case EStructProj(field @ _, fieldIndex @_, struct) =>
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/ExprTraversable.scala:30]:
```suggestion
      case EStructProj(field @ _, struct) =>
```

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala [hunk 5]
```diff
@@ -79,7 +79,7 @@ private[validation] object TypeTraversable {
       case EScenario(s) =>
         foreach(s, f)
       case EVar(_) | EVal(_) | EBuiltin(_) | EPrimCon(_) | EPrimLit(_) | EApp(_, _) | ECase(_, _) |
-          ELocation(_, _) | EStructCon(_) | EStructProj(_, _) | EStructUpd(_, _, _) |
+          ELocation(_, _) | EStructCon(_) | EStructProj(_, _, _) | EStructUpd(_, _, _) |
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala:82]:
```suggestion
          ELocation(_, _) | EStructCon(_) | EStructProj(_, _) | EStructUpd(_, _, _) |
```

**Review by remyhaemmerle-da — APPROVED** [review]
Not really fan of it, but I do not see other reasonable way to do it. 

I any case, I would really like the new filed to be internalized as suggested in the PR.


Furthermore no reason to expose that outside the type checker/compiler

**Comment by hurryabit** [comment]
FYI @nmarton-da

**Pull request merged**